src/compiler/checker.ts

```diff
@@ -13540,6 +13540,9 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         result.compositeSignatures = concatenate(left.compositeKind !== TypeFlags.Intersection && left.compositeSignatures || [left], [right]);
         if (paramMapper) {

```

Comment:
the gist of the issue was that first 2 calls with type params were unionized here (so `paramMapper` was created and attached to became `left.mapper` in the next iteration) but then that composite signature was unionized with one without the type params so the `left.mapper` wasn't attached to the `result` (because `paramMapper` was not created this time)